source/opt/allocator.h

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SOURCE_OPT_ALLOCATOR_H_
+#define SOURCE_OPT_ALLOCATOR_H_
+
+#include <cassert>
+#include <cstdlib>
+#include <limits>
+#include <map>
+#include <memory>
+#include <set>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace spvtools {
+
+// A batch allocator.
+//
+// This allocator wraps around malloc/free and conducts memory allocation on
+// a |kPageSize|-byte footprint. This groups multiple small allocation into a
+// larger one so that we can improve performance.
+//
+// This allocator adopts an ever-increasing approach; new memory are kept
+// allocated even if previously requested memory are deallocated. It's only
+// at the time of the destruction of this allocator that all the memory gets
+// released.
```

Comment:
How does our memory consumption change?  We might need to be careful how much we use.  There are other consideration when you start to use too much memory.  Are we in an environment that does not allow us to use that much memory?  I have seen bugs from the fuzzer because of stuff like this.  It can effect how many parallel compiles can be done if the machines has limited space.  There are also performance issues because of a lack of memory locality.

This type of allocation is not always a win.

Reply:
The way tls containers work is that the allocator's lifetime is tied to the container they are allocating for. So if the container gets actually destructed then the memory will get freed at that point.

However if the container is just cleared then the unused memory will accumulate.

Reply:
The STL container will have an instance of `StlAllocator`.  When that gets destroyed it will not free memory.  It will only get freed when `AllocatorImpl` is destroyed, which is owned by the `Optimizer`.  None of this memory will be deleted until that object is destroyed.